### DISCUSSION TITLE
A classification algorithm to segment & color code bainite and polygonal ferrite fractions

### DISCUSSION THREAD

**Discussion by Mikewangkang** [AzdiarGazder/mtexTools#5] [Q&A] [answered]
Hi @AzdiarGazder 

The developed Matlab®code uses the MTEX toolbox [33] to define first the grain boundaries according to the selected grain tolerance angle and afterwards calculate the KAM of each pixel. The program classifies grains as granular bainite if one pixel exceeds a misorientation of 3°; otherwise it is categorized as polygonal ferrite. Subsequently, the classification algorithm color-codes the classified grains according to their determined phase and calculates the measured granular bainite and polygonal fraction.I can calculate the KAM value for each pixel, but then the classification algorithm color codes the classified grains based on the determined phase and calculates the measured grain bainitic and polygonal fractions I don't know how to do that with code.As shown in the figure below, the left is what I can achieve currently, and the right is the drawing I want to achieve.I would appreciate it if you could help me.

![1693968637106](https://github.com/AzdiarGazder/mtexTools/assets/139696590/20d8f274-4636-4bff-92d8-74734e642918)

```matlab
%% Import Script for EBSD Data
%
% This script was automatically created by the import wizard. You should
% run the whoole script or parts of it in order to import your data. There
% is no problem in making any changes to this script.

%% Specify Crystal and Specimen Symmetries

% crystal symmetry
CS = {... 
  'notIndexed',...
  crystalSymmetry('m-3m', [2.9 2.9 2.9], 'mineral', 'Iron bcc (old)', 'color', [0.53 0.81 0.98]),...
  crystalSymmetry('m-3m', [3.7 3.7 3.7], 'mineral', 'Iron fcc', 'color', [0.56 0.74 0.56])};

% plotting convention
setMTEXpref('xAxisDirection','east');
setMTEXpref('zAxisDirection','intoPlane');

%% Specify File Names

% path to files
pname = 'D:\桌面\B钢数据\B-750\B钢\B钢EBSD\B750';

% which files to be imported
fname = [pname '\B750-1.crc'];

%% Import the Data

% create an EBSD variable containing the data
ebsd = EBSD.load(fname,CS,'interface','crc',...
  'convertEuler2SpatialReferenceFrame');

% The second parameter that is involved in grain reconstruction is the threshold misorientation angle indicating a grain boundary.
grains = calcGrains(ebsd,'angle',5*degree)

[grains,ebsd.grainId] = calcGrains(ebsd('indexed'));

% remove one-three pixel grains
ebsd(grains(grains.grainSize <= 3)) = [];
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'));

grains = smooth(grains,5);

plot(ebsd('indexed'),ebsd('indexed').orientations)
hold on
plot(grains.boundary,'lineWidth',1.5)
hold off

ebsd = ebsd.gridify;

kam = ebsd.KAM / degree;

% lets plot it
plot(ebsd,kam,'micronbar','off')
caxis([0,15])
mtexColorbar
mtexColorMap LaboTeX
hold on
plot(grains.boundary,'lineWidth',1.5)
hold off

plot(ebsd,ebsd.KAM('threshold',2.5*degree) ./ degree,'micronbar','off')
caxis([0,5])
mtexColorbar
mtexColorMap LaboTeX
hold on
plot(grains.boundary,'lineWidth',1.5)
hold off

% chose a denoising filter
F = halfQuadraticFilter;
F.alpha = 0.5;

% denoise the orientation map
ebsdS = smooth(ebsd,F,'fill',grains);

% plot the first order KAM
plot(ebsdS,ebsdS.KAM('threshold',2.5*degree) ./ degree,'micronbar','off')
caxis([0,5])
mtexColorbar
mtexColorMap LaboTeX
hold on
plot(grains.boundary,'lineWidth',1.5)
hold off
```

Best wishes
Wang Kang

**Comment by AzdiarGazder** [comment]
Hi @Mikewangkang,

Thank you for sharing your script.

There are a few issues that I'd like to clarify first:

1. As I mentioned in my last message, there are various methods used to classify and segment ferrite constituents. 
- One method, which I developed, can be found [here](https://www.sciencedirect.com/science/article/pii/S0304399114001284).
You can apply my method by going through the various steps listed in the paper.
- Another method, developed by Sachin Shrestha, can be found [here](https://www.sciencedirect.com/science/article/pii/S0304399113002957).
I have modified Sachin's method to work within MTEX [here](https://github.com/AzdiarGazder/mtexTools/tree/main/ferriteQuantifier).

Have you applied either of these two methods to your map(s)?
If not, I strongly encourage you to please use either of these methods.
If you have already used them, please let me know if/why they did not work in your case.



2. I'd like to reiterate the issue with using KAM:
- Based on the information you've provided, you seem to suggest that 1 criterion (like KAM) is enough to segment between constituents. However, modern segmentation methods, like the two I've shown above, use multiple criteria to separate between the constituents. Using only 1 criterion is simply not enough to accurately (and consistently) segment between constituents.
- A criterion like KAM is a pixel-based criterion. When segmenting constituents, it is advisable to use grain-based criteria like GAM and GOS. This enables cleaner and easier segmentation.
- Depending on the thermo-mechanical history of each individual microstructure (and consequently, the map data), the thresholds should be adjusted/changed to accurately (and consistently) segment between constituents. Using 1 threshold value across all microstructures and maps is again simply not correct.



3. Based on my understanding of your script, you are removing grains smaller than 3 pixels, smoothing grain boundaries, and then plotting KAM maps: (i) before filtering, (ii) before filtering and after applying a threshold, and (iii) after filtering. 
- You have not undertaken any segmentation of the microstructure as yet. 
- You are already using smoothing and denoising filters in your script.
This which affect any segmentation being undertaken using a pixel based criterion like KAM.
This is one of the reasons why multiple criteria are required for successful segmentation.

Warm regards,
Azdi

**Reply by Mikewangkang** [reply]
Hi！ @AzdiarGazder

Thank you for your reply. My purpose is to conduct a simple two-phase distinction and then count the two-phase fraction for modeling. However, the KAM values of my sample with granular bainite and polygonal ferrite are very different. 

I saw an article uses KAM for two-phase distinction, so I thought about trying this method. 

Then I tried using your code based on Sachin Shrestha's method, and it gave an error with my EBSD data. The hint needs to be single phase, while my data has two phases in addition to the unindexed ones. I still get an error after adding the code specifying single phase, is there any way you can improve it? I only want the 'Iron bcc (old)' phase. The picture shows the code I am running and the error message. 

As for your method, it is perfect, but manual statistics are required. If different EBSD data are used for statistics in the later stage, I think it will consume a lot of time.


Regarding my code, I just found the KAM calculation method from the mtex official website and edited it with reference.

![5fa7b38dd78ac6d19fa730beb5c57fe](https://github.com/AzdiarGazder/mtexTools/assets/139696590/5efe0514-0936-4f60-ad2a-ee77aa2404d1)

![d1358c22c3d974ed463d71e7772e5cf](https://github.com/AzdiarGazder/mtexTools/assets/139696590/90daa868-6a95-4273-ba61-347a4e1ce960)

![899cc2690644ac0635465052b2acdeb](https://github.com/AzdiarGazder/mtexTools/assets/139696590/2284d7b7-12cd-4012-98ba-df131bc8645e)

Best wishes
Wang Kang

**Reply by AzdiarGazder** [reply, marked as answer]
Hi @Mikewangkang,

1. Thank you for pointing out the limitation of the original ferrite quantifier. I have added a **[second script to the folder](https://github.com/AzdiarGazder/mtexTools/tree/main/ferriteQuantifier)** that deals specifically with multi-phase maps. I have also added a example map of duplex steel so that you can run the script and see how it works without making any changes. **If this resolves your issue, please mark this message as your answer.**

2. I would be grateful if you could please send me a link to the article you are citing that uses the KAM value. 

Warm regards,
Azdi

**Reply by Mikewangkang** [reply]
Hi @AzdiarGazder 
Thank you very much for your help, I will try the new code and if it works I will mark it as my answer. Given below is the article I quoted. [Jentner et al., Automated classification of granular bainite and polygonal ferrite by EBSD.](https://github.com/AzdiarGazder/mtexTools/files/12565238/Automated.classification.of.granular.bainite.and.polygonal.ferrite.by.electron.pdf)
Best wishes
Wangkang
